### PR TITLE
feat: Project Option Epochs

### PR DESCRIPTION
**Pull request by mitsuhiko**
This adds epoch based defaults support to project options. When a project is created the current epoch is frozen and defaults are resolved that were latest at the time of the project creation. Whenever new defaults should be introduced the epoch is incremented.

This lets us opt new projects into newer JS loader versions or grouping configs.

### CODE REVIEW

**Comment by mitsuhiko** [comment]
The idea is that new projects freeze in the latest epoch so we can do epoch specific defaults (eg: projects created after X have newer loader default, grouping logic default etc.)

File: src/sentry/projectoptions/manager.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+            try:
+                return self.epoch_defaults[idx][1]
+            except IndexError:
+                pass
+        return self.default
+
+
+class ProjectOptionsManager(object):
+    """Project options used to be implemented in a relatively ad-hoc manner
+    in the past.  The project manager still uses the functionality of the
```
Comment (markstory) [inline src/sentry/projectoptions/manager.py]: This sentence seems incomplete.

**Review by markstory — COMMENTED** [review]
Could the additional queries used to resolve the epoch defaults impact event processing in a negative way?

**Comment by mitsuhiko** [comment]
@markstory we pre-fetch all options in one go for a project and the epoch is stored as an option. Should thus not have a measurable impact.

**Comment by markstory** [comment]
Would it be simpler if we baked the current epoch'a defaults into a project's options when it is created going forward? That would allow us to evolve the defaults for future epochs without having to maintain it in code. We would still need a fallback for the pre-baked option epoch.

This would be contingent on users not being able to erase these load bearing options.

**Comment by mitsuhiko** [comment]
That would require massive migrations every time an option is changed. Also most projects are probably relatively dormant to inactive which would bloat the database very quickly.

**Comment by markstory** [comment]
@mitsuhiko Why would we need migrations? I'm suggesting that when a new project is created we save all of the current epoch's option defaults onto the project. Each project going forward would have its default values set. As we evolve the defaults new projects would continue to be minted with the current defaults. Projects with no options set would fallback to the v1 defaults which exist in code.

**Comment by mitsuhiko** [comment]
@markstory fair, but i'm not sure it's a great idea. right now on average we have 5.9 options stored per project. This will bump this up to 10 immediately (we plan on adding 4 epoch bound options for a start). That might not be a huge concern but it seems suboptimal to have to store default values when we can calculate them on the fly.

**Comment by HazAT** [comment]
Considering the loader doesn't use project_options, rather it uses `project_key.data.get('browserSdkVersion', DEFAULT_VERSION)` so I just wanted to point out that this doesn't solve it for the loader.

**Review by HazAT — APPROVED** [review]
LGTM, cool feature. 
We could also add it for reprocessing but maybe not in this PR since it's already quite big.

**Review by markstory — APPROVED** [review]

**Pull request merged**